### ISSUE TITLE
Questions about MM3D mesh import.

### ISSUE THREAD

**Issue by JamesO2x** [MeltyPlayer/OoT3D-Importer#6]
Hi. I have a few questions about importing MM3D scene models. 

I noticed that MM3D isn't checked off on the "To Do" section of your readme, so I assume that means "its not 100% compatible yet." But since MM3D uses a pretty similar CMB format, I've found that **many** of the MM3D CMB files do import correctly with no issues. Awesome!

### Question 1:
I know that in order to convert a ZSI file into a CMB, I have to use a hex editor to remove the ZSI header part from the file. I can do this myself, no issue. So I was wondering if: is there any specific reason that this addon itself can't read the ZSI directly, and simply skip over the bytes for the ZSI header by searching for the CMB header part? Is that planned for the future?

If not, do you know of any tool (a program or python script, etc) that can automatically remove the ZSI header from multiple files in bulk, so I that I don't have to do this manually?

If not, I suppose I can figure out a python script myself that would be able to do this.

### Question 2:
I've noticed that some of the ZSI files forMM3D have an additional header before the ZSI section that starts with "LzS". I did a bit of research and my best guess is that this has something to do with "LZSS compression", as the data in the files also appears to be compressed. Likewise, I noticed that this addon does not seem to be able to read these ZSI/CMB files, I'm guessing because of they byte compression.

Is there any plan for supporting these files?

If not, do you know of any tool that can decode the LZSS compression (assuming I'm right about this) and extract the ZSI file in an uncompressed state?

I noticed the [N3DSCmbViewer](https://github.com/xdanieldzd/N3DSCmbViewer) is able to open these compressed ZSI files directly. But my issue with that tool is that it doesn't seem possible to then extract the CMB from the ZSI once loaded (it appears to be disabled for Scene files). Also, the N3DSCmbViewer has Collada export, but the resulting DAE file has some issues with UV accuracy, and the repo is also no longer being maintained.

### Question 3:
Are there any plans for an updated 2.80+ version of this addon? I know the Blender API had a pretty big rewrite, as well as an overhaul with the material system. So I don't expect you to do a rewrite of the addon. A simple "yes" or "no" answer on this would be nice though.

So far I've found that importing into 2.79, saving the Blend file, and re-opening that Blend in 3.0, seems to work just fine. No data appears to be lost.

---

In terms of accuracy with ripping the meshes, your addon seems to be perfect -- correct UVs, correct Vertex RGBA, correct bone structure, etc. The only major issue is that I can't seem to find a way to import these compressed files.

If my assumptions about LZSS compression are correct, then I might be able to figure out a solution using Python. But I want to confirm that is the issue before I sink dozens of hours into a solution that's incorrect. So any comment you could make on this would be helpful. Thanks.

**Comment by MeltyPlayer** [comment]
Thanks for filing an issue, and apologies for the delayed response.

Unfortunately, I ran into a number of issues trying to develop this plugin (lack of Python type safety, slow development speed due to needing to restart Blender each time, difficulty mapping the model’s bone axes to the unique orientation used by Blender). So I decided to pivot towards implementing everything in my new C# tool instead, [FinModelUtility](https://github.com/MeltyPlayer/FinModelUtility), which doesn’t have any of those issues. 

This should be better in the long run, as it should be much quicker to develop this way.

But in regards to your questions specifically:

1. I don’t remember the full reasoning behind this, but I believe it was because ZSI files can contain multiple CMB models, which can lead to ambiguity issues—which model should be loaded specifically? This is something that was resolved in the FinModelUtility, which automatically extracts the contents of ZSI files. I don’t know of a tool for automatically removing these headers though, unfortunately.
2. I wasn’t able to find anything else for handling LZSS compression, but I was able to port this logic you linked to add support for automatically decompressing these files in the FinModelUtility. (I still need to add a .bat file for Majora’s Mask 3D though, oops.) I believe you should then be able to use the Blender plugin with these decompressed files, but correct me if I’m wrong.
3. I don’t plan to migrate this to new versions of Blender due to the aforementioned development issues, unfortunately.

**Comment by JamesO2x** [comment]
Thanks for the comment. I tested the FinModelUtility, it seems like a pretty robust solution for all sorts of general purpose ripping. Awesome work!

I tested with Ocarina of Time 3D, but I noticed the Utility only seems to rip actor models and not scene models. I posted a comment about it here: https://github.com/MeltyPlayer/FinModelUtility/issues/1#issuecomment-1169336501

I don't know if this is a limitation of the Utility or if I did something wrong during the setup. At the moment, I am only interested in ripping scene models from OoT3D and MM3D.

Unfortunately I can't do any further testing at the moment, as I'm struggling with carpal tunnel syndrome. I can only write here somewhat thanks to microphone dictation. But general computer use is difficult. So please forgive me if it takes me awhile to respond.

**Comment by MeltyPlayer** [comment]
Yes, the OoT3D implementation in the FinModelUtility is still a work-in-progress.

The core issue with OoT3D is one of the steps in the model ripping process, grouping model files together. Each model uses some set of model/animation/texture files from the ROM. Unfortunately, which models/animations/textures go together doesn’t seem to be defined anywhere in the ROM, and the naming for files is somewhat haphazard, so these often need to be manually grouped together by a human. Making this full list of models is a tedious process—I’ve only partially done it for actors, and haven’t dug into scenes yet.

I believe that this Blender plugin should work for scenes in the meantime, but yes, I’ll need to look into handling scenes at some point.

**Comment by JamesO2x** [comment]
Thanks for the info.

For MM3D I made a list of all the scene files, with their corresponding in-game names. And for areas that don't have an in-game name, I gave them a plain-english name that makes sense. I also grouped them all by whether they are "interior" rooms or "exterior" scenes, or part of a dungeon or sub-dungeon (like the Pirate Fortress).

https://docs.google.com/spreadsheets/d/1GJhEUT6deJRxYG2wpkITJwb_N6XZX4v57w6qQuxv6Zc/edit?usp=sharing

Some of the columns are just notes for my own purposes, but this list should save a bit of time figuring out what's what.

I haven't made a list like this for OoT3D, so just MM3D for now.
